File: projects/native-apps/ios/App/LaunchIntro.swift
import Foundation

enum LaunchIntroEnd: Equatable {
	case playedToEnd
	case assetFailed
	case timedOut
	case skipped
}

enum LaunchIntroPhase: Equatable {
	case idle
	case playing
	case fading
	case finished
}

struct LaunchIntroOverlay: Equatable {
	let showsVideo: Bool
	let opacity: Double
	let usesDarkBackdrop: Bool
}

enum LaunchIntro {
	static func initialPhase(isFirstLaunch: Bool, reduceMotion: Bool) -> LaunchIntroPhase {
		isFirstLaunch && !reduceMotion ? .playing : .idle
	}

	static func overlay(for phase: LaunchIntroPhase) -> LaunchIntroOverlay {
		switch phase {
		case .idle, .finished:
			return LaunchIntroOverlay(showsVideo: false, opacity: 0, usesDarkBackdrop: false)
		case .playing:
			return LaunchIntroOverlay(showsVideo: true, opacity: 1, usesDarkBackdrop: true)
		case .fading:
			return LaunchIntroOverlay(showsVideo: true, opacity: 0, usesDarkBackdrop: false)
		}
	}

	static func next(after phase: LaunchIntroPhase, end: LaunchIntroEnd) -> LaunchIntroPhase {
		guard phase == .playing else { return phase }
		return end == .skipped ? .finished : .fading
	}

	static func playsMusic(isLoggedIn: Bool, isForeground: Bool) -> Bool {
		!isLoggedIn && isForeground
	}

	static let fadeDuration = 0.45

	static let watchdogSlack = 3.0

	static let videoDuration = 8.083333

	static let logoScreenFraction = 0.378
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/app/LaunchIntro.kt
package com.readplace.android.app

enum class LaunchIntroEnd {
	PLAYED_TO_END,
	ASSET_FAILED,
	TIMED_OUT,
	SKIPPED,
}

enum class LaunchIntroPhase {
	IDLE,
	PLAYING,
	FADING,
	FINISHED,
}

data class LaunchIntroOverlay(
	val showsVideo: Boolean,
	val opacity: Double,
	val usesDarkBackdrop: Boolean,
)

object LaunchIntro {
	fun initialPhase(isFirstLaunch: Boolean, reduceMotion: Boolean): LaunchIntroPhase =
		if (isFirstLaunch && !reduceMotion) LaunchIntroPhase.PLAYING else LaunchIntroPhase.IDLE

	fun overlay(phase: LaunchIntroPhase): LaunchIntroOverlay =
		when (phase) {
			LaunchIntroPhase.IDLE, LaunchIntroPhase.FINISHED ->
				LaunchIntroOverlay(showsVideo = false, opacity = 0.0, usesDarkBackdrop = false)
			LaunchIntroPhase.PLAYING ->
				LaunchIntroOverlay(showsVideo = true, opacity = 1.0, usesDarkBackdrop = true)
			LaunchIntroPhase.FADING ->
				LaunchIntroOverlay(showsVideo = true, opacity = 0.0, usesDarkBackdrop = false)
		}

	fun next(after: LaunchIntroPhase, end: LaunchIntroEnd): LaunchIntroPhase {
		if (after != LaunchIntroPhase.PLAYING) return after
		return if (end == LaunchIntroEnd.SKIPPED) LaunchIntroPhase.FINISHED else LaunchIntroPhase.FADING
	}

	fun playsMusic(isLoggedIn: Boolean, isForeground: Boolean): Boolean = !isLoggedIn && isForeground

	const val FADE_DURATION = 0.45

	const val WATCHDOG_SLACK = 3.0

	const val VIDEO_DURATION = 8.083333

	const val LOGO_SCREEN_FRACTION = 0.378
}
